File: services/edge-node/src/state.ts
// Local state file for the Edge Node.
//
// Stores the dpfedge_* node token + edgeNodeId + Authority-decided
// intervals so the agent doesn't have to re-enroll on every restart.
//
// Phase 0 storage: plain JSON at $DPF_EDGE_STATE_DIR/state.json with
// 0600 perms. The spec calls out OS-secure-store (Keychain, Credential
// Manager, libsecret) as preferred, but in the Linux-container Phase 0
// deployment the host's libsecret isn't available. The 0600 file under
// a container-owned directory is the agreed Phase 0 fallback per the
// spec § Phase 0 storage downgrade (Linux container Mode 1).
//
// Read-time enforcement per spec:
//   - File mode must be 0600 (refuse if loosened by host bind-mount,
//     volume export, or operator copy).
//   - File owner UID must match the current process UID (refuse if
//     a different account dropped state into our directory — e.g. a
//     host bind-mount leaking a different container's state).
// Cross-container fingerprint detection (the spec's stronger
// promise) is Phase 1+; for now the docker-managed-volume + non-root
// dedicated UID controls bound the risk.

import { promises as fs } from "node:fs";
import { dirname, join } from "node:path";
import * as os from "node:os";

import { z } from "zod";

const StateSchema = z.object({
  /** Stable nodeId returned by the Authority at enrollment. */
  nodeId: z.string().min(1),
  /** Plaintext dpfedge_* node token. NEVER logged. */
  nodeToken: z.string().min(1),
  /** When this state was last written. */
  enrolledAt: z.string().datetime(),
  /** Authority-decided heartbeat interval in seconds. */
  heartbeatIntervalSec: z.number().int().positive(),
  /** Authority-decided sweep interval in seconds. */
  sweepIntervalSec: z.number().int().positive(),
  /** Authority-decided metrics collection interval in seconds. Defaults to 10 s. */
  metricsIntervalSec: z.number().int().positive().optional(),
  /** Capabilities the Authority accepts from this node. */
  acceptedCapabilities: z.array(z.string()),
  /** trustState at last heartbeat. */
  trustState: z.enum(["pending", "trusted", "quarantined", "revoked"]),
});

export type EdgeNodeState = z.infer<typeof StateSchema>;

const STATE_FILENAME = "state.json";

export function statePath(stateDir: string): string {
  return join(stateDir, STATE_FILENAME);
}

/**
 * Get the current process UID. Returns null on platforms without
 * POSIX UIDs (Windows in tests). Owner enforcement is skipped on
 * those platforms; the spec's Mode 1 controls only bind in the
 * Linux-container deployment where UIDs are real.
 */
function currentUid(): number | null {
  if (typeof process.getuid !== "function") return null;
  try {
    return process.getuid();
  } catch {
    return null;
  }
}

/**
 * Skip the file-mode and owner checks on hosts that don't have POSIX
 * permission semantics (Windows). The container deployment is Linux;
 * skipping on Windows preserves dev-host ergonomics while keeping the
 * production-path checks intact.
 */
function shouldEnforcePosixPerms(): boolean {
  return os.platform() !== "win32";
}

/**
 * Verify the state file's mode is 0600 and owner is the current
 * process UID. Throws with a descriptive error if either check fails
 * — refusing to read state we don't trust is the point.
 */
async function verifyStatePerms(path: string): Promise<void> {
  if (!shouldEnforcePosixPerms()) return;

  const stat = await fs.stat(path);

  // Mode check: the lower 9 bits encode rwxrwxrwx. We require 0600
  // exactly (owner read/write, no group, no other). If the file is
  // group/world readable, the token is exposed to anything else
  // running as that group or on that host.
  const modeBits = stat.mode & 0o777;
  if (modeBits !== 0o600) {
    throw new Error(
      `Edge Node state file at ${path} has unsafe permissions: `
      + `expected mode 0600, got 0${modeBits.toString(8).padStart(3, "0")}. `
      + `Refusing to load; the node token in this file may be exposed. `
      + `Fix: chmod 600 ${path} (and investigate how the loose mode got there).`,
    );
  }

  // Owner check: the file must be owned by the current process UID.
  // A different owner means either (a) a host bind-mount leaked
  // another account's state into this container, or (b) the
  // installer ran as a different account than the runtime. Either
  // way, the token in the file is not ours to use.
  const myUid = currentUid();
  if (myUid !== null && stat.uid !== myUid) {
    throw new Error(
      `Edge Node state file at ${path} is owned by UID ${stat.uid}, `
      + `but the current process runs as UID ${myUid}. `
      + `Refusing to load; cross-account state-file read is the host-bind-mount-leak threat the spec's storage downgrade controls bound. `
      + `Fix: ensure the state directory is owned by the runtime's dedicated UID (e.g. dpf:dpf in the Dockerfile), and the state directory is a docker-managed volume, not a host bind-mount.`,
    );
  }
}

/**
 * Read state from disk. Returns null if the state file doesn't exist
 * (caller should treat as "first run" and trigger enrollment). Throws
 * if:
 *   - the file exists but parsing fails (corruption shouldn't be
 *     silently ignored);
 *   - the file mode is not 0600 (per spec § Phase 0 storage downgrade);
 *   - the file owner UID doesn't match the current process UID (same).
 */
export async function loadState(stateDir: string): Promise<EdgeNodeState | null> {
  const path = statePath(stateDir);

  // Existence check first via stat — lets us return null cleanly for
  // ENOENT without doing the perm check on a phantom path. Any other
  // stat error (EACCES, EPERM, etc.) propagates.
  let exists = true;
  try {
    await fs.access(path);
  } catch (err: unknown) {
    if (
      typeof err === "object" &&
      err !== null &&
      "code" in err &&
      (err as { code?: string }).code === "ENOENT"
    ) {
      exists = false;
    } else {
      throw err;
    }
  }
  if (!exists) return null;

  // File exists — enforce the mode + owner controls BEFORE reading
  // the contents. Reading first would mean a loose-perms file gets
  // its plaintext token slurped into the Node process anyway.
  await verifyStatePerms(path);

  const raw = await fs.readFile(path, "utf8");

  let parsed: unknown;
  try {
    parsed = JSON.parse(raw);
  } catch (err) {
    throw new Error(`Edge Node state file at ${path} is corrupt: ${(err as Error).message}`);
  }

  const result = StateSchema.safeParse(parsed);
  if (!result.success) {
    const issues = result.error.issues
      .map((i) => `  - ${i.path.join(".")}: ${i.message}`)
      .join("\n");
    throw new Error(
      `Edge Node state file at ${path} is missing required fields:\n${issues}`,
    );
  }
  return result.data;
}

/**
 * Atomically write state. Uses tmp-file + rename so a crash mid-write
 * leaves the previous state intact. Sets 0600 perms.
 */
export async function saveState(
  stateDir: string,
  state: EdgeNodeState,
): Promise<void> {
  // Validate before writing — we don't want to persist an invalid
  // state shape that would fail to load on next restart.
  StateSchema.parse(state);

  await fs.mkdir(dirname(statePath(stateDir)), { recursive: true });
  await fs.mkdir(stateDir, { recursive: true });

  const path = statePath(stateDir);
  const tmp = `${path}.tmp.${process.pid}`;
  const body = `${JSON.stringify(state, null, 2)}\n`;

  await fs.writeFile(tmp, body, { mode: 0o600 });
  await fs.rename(tmp, path);
  // Re-chmod after rename in case the umask shifted bits.
  await fs.chmod(path, 0o600);
}

/**
 * Clear state entirely (e.g. on revocation). The next start triggers
 * a fresh enrollment.
 */
export async function clearState(stateDir: string): Promise<void> {
  const path = statePath(stateDir);
  try {
    await fs.unlink(path);
  } catch (err: unknown) {
    if (
      typeof err === "object" &&
      err !== null &&
      "code" in err &&
      (err as { code?: string }).code === "ENOENT"
    ) {
      return;
    }
    throw err;
  }
}

File: services/edge-node-go/internal/state/state.go
// Package state persists the Edge Node's enrollment record (nodeId +
// nodeToken + Authority-decided intervals + trustState) across restarts.
// Mirrors services/edge-node/src/state.ts byte-for-byte on the wire so
// either runtime can read state written by the other (wire-contract
// parity).
//
// Phase 0 storage: plain JSON at $DPF_EDGE_STATE_DIR/state.json with
// 0600 perms. OS-secure-store (Keychain, Credential Manager, libsecret)
// is the Phase-1+ target; for now the file-system controls (mode,
// owner) bound the risk.
//
// Read-time enforcement (POSIX hosts only — skipped on Windows where
// the TS implementation also skips it; both runtimes apply the
// equivalent ACL-based controls in Phase 1):
//   - File mode must be 0600. A loosened mode means the token in the
//     file may be exposed to anything running as the same group/world.
//   - File owner UID must match the current process UID. A different
//     owner means either a host bind-mount leaked another account's
//     state or the installer ran as a different account than the
//     runtime — in either case, the token isn't ours to use.
package state

import (
	"encoding/json"
	"errors"
	"fmt"
	"io/fs"
	"os"
	"path/filepath"
)

// EdgeNodeState mirrors the TypeScript StateSchema exactly. JSON tags
// must match so either runtime can read either's state file.
type EdgeNodeState struct {
	NodeID               string   `json:"nodeId"`
	NodeToken            string   `json:"nodeToken"`
	EnrolledAt           string   `json:"enrolledAt"` // RFC 3339 / ISO 8601
	HeartbeatIntervalSec int      `json:"heartbeatIntervalSec"`
	SweepIntervalSec     int      `json:"sweepIntervalSec"`
	MetricsIntervalSec   *int     `json:"metricsIntervalSec,omitempty"`
	AcceptedCapabilities []string `json:"acceptedCapabilities"`
	TrustState           string   `json:"trustState"` // pending | trusted | quarantined | revoked
}

const stateFilename = "state.json"

// Path returns the on-disk location of the state file.
func Path(stateDir string) string {
	return filepath.Join(stateDir, stateFilename)
}

// Load reads state from disk. Returns (nil, nil) when the file doesn't
// exist — the caller treats that as "first run" and triggers
// enrollment. Returns an error when the file exists but is corrupt,
// has the wrong mode, or has the wrong owner.
func Load(stateDir string) (*EdgeNodeState, error) {
	path := Path(stateDir)

	info, err := os.Stat(path)
	if err != nil {
		if errors.Is(err, fs.ErrNotExist) {
			return nil, nil
		}
		return nil, fmt.Errorf("stat %s: %w", path, err)
	}

	if err := verifyPosixPerms(path, info); err != nil {
		return nil, err
	}

	raw, err := os.ReadFile(path)
	if err != nil {
		return nil, fmt.Errorf("read %s: %w", path, err)
	}

	var s EdgeNodeState
	if err := json.Unmarshal(raw, &s); err != nil {
		return nil, fmt.Errorf("state file %s is corrupt: %w", path, err)
	}
	if err := validate(&s); err != nil {
		return nil, fmt.Errorf("state file %s is missing required fields: %w", path, err)
	}
	return &s, nil
}

// Save atomically writes state to $stateDir/state.json with 0600 perms.
// Uses a tmp-file + rename so a crash mid-write leaves the previous
// state intact.
func Save(stateDir string, s *EdgeNodeState) error {
	if err := validate(s); err != nil {
		return fmt.Errorf("validate state before save: %w", err)
	}
	if err := os.MkdirAll(stateDir, 0o700); err != nil {
		return fmt.Errorf("mkdir %s: %w", stateDir, err)
	}

	body, err := json.MarshalIndent(s, "", "  ")
	if err != nil {
		return fmt.Errorf("marshal state: %w", err)
	}
	body = append(body, '\n')

	path := Path(stateDir)
	tmp := fmt.Sprintf("%s.tmp.%d", path, os.Getpid())

	// O_EXCL on the tmp file so two concurrent agents don't fight over
	// the same slot. If this fails because the tmp file already exists,
	// the operator can `rm` it; that's a rare recovery path.
	f, err := os.OpenFile(tmp, os.O_WRONLY|os.O_CREATE|os.O_EXCL|os.O_TRUNC, 0o600)
	if err != nil {
		return fmt.Errorf("open %s: %w", tmp, err)
	}
	if _, err := f.Write(body); err != nil {
		_ = f.Close()
		_ = os.Remove(tmp)
		return fmt.Errorf("write %s: %w", tmp, err)
	}
	if err := f.Close(); err != nil {
		_ = os.Remove(tmp)
		return fmt.Errorf("close %s: %w", tmp, err)
	}

	if err := os.Rename(tmp, path); err != nil {
		_ = os.Remove(tmp)
		return fmt.Errorf("rename %s -> %s: %w", tmp, path, err)
	}

	// Re-chmod after rename in case the umask shifted bits. Best
	// effort on Windows where mode bits beyond the read-only attribute
	// are advisory.
	if shouldEnforcePosixPerms() {
		if err := os.Chmod(path, 0o600); err != nil {
			return fmt.Errorf("chmod %s: %w", path, err)
		}
	}
	return nil
}

// Clear deletes the state file. ENOENT is treated as success; the next
// start will trigger fresh enrollment.
func Clear(stateDir string) error {
	err := os.Remove(Path(stateDir))
	if err != nil && !errors.Is(err, fs.ErrNotExist) {
		return err
	}
	return nil
}

// validate enforces the same shape constraints the TypeScript
// StateSchema enforces. Any drift between this list and state.ts is a
// wire-contract regression — the parity tests catch it.
func validate(s *EdgeNodeState) error {
	if s == nil {
		return errors.New("state is nil")
	}
	if s.NodeID == "" {
		return errors.New("nodeId is required")
	}
	if s.NodeToken == "" {
		return errors.New("nodeToken is required")
	}
	if s.EnrolledAt == "" {
		return errors.New("enrolledAt is required")
	}
	if s.HeartbeatIntervalSec <= 0 {
		return errors.New("heartbeatIntervalSec must be a positive integer")
	}
	if s.SweepIntervalSec <= 0 {
		return errors.New("sweepIntervalSec must be a positive integer")
	}
	if s.MetricsIntervalSec != nil && *s.MetricsIntervalSec <= 0 {
		return errors.New("metricsIntervalSec, if set, must be a positive integer")
	}
	switch s.TrustState {
	case "pending", "trusted", "quarantined", "revoked":
	default:
		return fmt.Errorf("trustState %q is not one of pending|trusted|quarantined|revoked", s.TrustState)
	}
	if s.AcceptedCapabilities == nil {
		return errors.New("acceptedCapabilities is required (may be empty array)")
	}
	return nil
}
